### ISSUE TITLE
More complete throw/yield detection

### ISSUE THREAD

**Issue by brettz9** [gajus/eslint-plugin-jsdoc#682]
## Expected behavior

For our utility `hasThrowValue` (used by `require-throws`) and `hasYieldValue` (used by `require-yields` and `require-yields-check`), I believe we may need to support some of the following AST branches (even if not of interest to `hasReturns` on which these functions were based).

```js
  case 'SequenceExpression': // Comma
  case 'ArrayExpression': case 'AssignmentExpression': case 'AssignmentPattern':
  case 'AwaitExpression': case 'BinaryExpression': case 'ConditionalExpression':
  case 'LogicalExpression': case 'MemberExpression': case 'OptionalMemberExpression':
  case 'VariableDeclaration':
  case 'OptionalCallExpression':
  case 'TaggedTemplateExpression':
  case 'TemplateElement': case 'TemplateLiteral': case 'UnaryExpression':
  case 'SpreadElement':
  case 'ArrayPattern': case 'ObjectPattern':
  case 'ObjectExpression':
  case 'Property':
  case 'ClassProperty':
  case 'ClassDeclaration': case 'ClassExpression': case 'MethodDefinition':
  case 'Super':
  case 'ExportDefaultDeclaration': case 'ExportNamedDeclaration':
  case 'LabeledStatement':
  case 'Import':
  case 'ImportExpression':
  case 'Decorator':
  case 'YieldExpression':
```

And if #683 is accepted, these checks will be needed for `resolve` as well (in its `hasValueOrExecutorHasNonEmptyResolveValue`).

## Actual behavior

Treats those expressions (and anything within them, including any `throw`/`yield`) as not throwing/not yielding.

So, `require-throws` and `require-yields` may falsely fail to report and `require-yields-check` can report when it should not.

## ESLint Config

```js
{
  'jsdoc/require-throws': 'error',
  'jsdoc/require-yields': 'error',
  'jsdoc/require-yields-check': 'error',
}
```

## ESLint sample

```js
/**
 *
 */
function * quux () {
  false || (() => {
    throw new Error();
  })();

  var a = [yield];
}
```

## Environment

- Node version: 12.16.2
- ESLint version  7.18.0
- `eslint-plugin-jsdoc` version: 31.4.0
